gateway/ln-gateway/src/actor_fork.rs

```diff
@@ -44,7 +55,112 @@ impl GatewayActor {
+                            warn!("Stream error: {:?}", e);
+                            None
+                        }
+                    } {
+                        // TODO: Assert short channel id matches the one we subscribed to, or cancel processing of intercepted HTLC
+                        // TODO: Assert the offered fee derived from invoice amount and outgoing amount is acceptable or cancel processing of intercepted HTLC
+                        // TODO: Assert the HTLC expiry or cancel processing of intercepted HTLC
+
+                        let hash = match sha256::Hash::from_slice(&payment_hash) {
+                            Ok(hash) => hash,
+                            Err(e) => {
+                                let fail = "Failed to parse payment hash";
+
+                                error!("{}: {:?}", fail, e);
+                                htlc_outcomes.push(CompleteHtlcsRequest {
+                                    action: Some(Action::Cancel(Cancel {
+                                        reason: fail.to_string(),
+                                        intercepted_htlc_id,
+                                    })),
+                                });
+                                continue;
+                            }
+                        };
+
+                        let amount_msat = Amount::from_msats(outgoing_amount_msat);
+
+                        let outcome = match actor.buy_preimage_internal(&hash, &amount_msat).await {
+                            Ok(preimage) => {
+                                info!("Successfully processed intercepted HTLC");
+                                CompleteHtlcsRequest {
+                                    action: Some(Action::Settle(Settle {
+                                        preimage: preimage.0.to_vec(),
+                                        intercepted_htlc_id,
+                                    })),
+                                }
+                            }
+                            Err(e) => {
+                                error!("Failed to process intercepted HTLC: {:?}", e);
+                                CompleteHtlcsRequest {
+                                    action: Some(Action::Cancel(Cancel {
+                                        reason: e.to_string(),
+                                        intercepted_htlc_id,
+                                    })),
+                                }
+                            }
+                        };
+
+                        htlc_outcomes.push(outcome);
+                    }
+
+                    if let Err(e) = lnrpc.complete_htlcs(htlc_outcomes).await {
+                        error!("Failed to complete HTLCs: {:?}", e);
+                        // NOTE: This is a potential loss of funds for the gateway.
+                        // We should consider a retry of this operation, cancel the HTLCs or reclaim funds.
+                    }
```

Comment:
This will only be hit once the stream closes, which also means that `complete_htlcs` may never be called and `htlc_outcomes` could grow indefinitely.

I already [mentioned this DoS vector in a previous iteration of the PR](https://github.com/fedimint/fedimint/pull/1198#discussion_r1069643531) and did not receive a rebuttal.

Reply:
> This will only be hit once the stream closes, which also means that complete_htlcs may never be called and htlc_outcomes could grow indefinitely.
> I already mentioned this DoS vector in a previous iteration of the PR and did not receive a rebuttal.

I agree there is a high severity logical bug here. Seems like we need additional markers at which we flush the `htlc_outcomes`, and also guarantee flushing of any contents just before we quit and close our subscription

[pending]

Reply:
I'd just call `complete_htlcs` after every successful buying of a preimage. We don't need batching for lack of expected load and additional latency might cost us actual money (30s timeout).

My proposal: get rid of outer loop, rename `complete_htlcs` to `complete_htlc` and make it take only one completion event instead of a vector, find a way to re-subscribe to HTLC events if the stream ever dies.

Also, what are the delivery guarantees if a stream dies because of e.g. an internet interruption? Do we have to resubmit all pending HTLCs?

Reply:
You know if I just got rid of this batching and streaming in all places except subscribe htlcs, my life would be way simpler!
On it!

Reply:
Nice